server/src/main/java/org/elasticsearch/threadpool/PrioFixedExecutorBuilder.java

```diff
@@ -0,0 +1,94 @@
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package org.elasticsearch.threadpool;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.PriorityBlockingQueue;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+
+import org.elasticsearch.common.Priority;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.SizeValue;
+import org.elasticsearch.common.util.concurrent.EsAbortPolicy;
+import org.elasticsearch.common.util.concurrent.EsExecutors;
+import org.elasticsearch.common.util.concurrent.EsThreadPoolExecutor;
+import org.elasticsearch.common.util.concurrent.PrioritizedRunnable;
+import org.elasticsearch.common.util.concurrent.PriorityRunnable;
+import org.elasticsearch.common.util.concurrent.SizeBlockingQueue;
+import org.elasticsearch.node.Node;
+
+/**
+ * Like {@link FixedExecutorBuilder} but uses a PriorityQueue. If scheduling
+ * {@link Runnable} that are not {@link PrioritizedRunnable} it wraps them into
+ * a {@link PriorityRunnable} with priority normal.
+ **/
+public final class PrioFixedExecutorBuilder extends FixedExecutorBuilder {
+
+    /**
+     * Construct a fixed executor builder.
+     *
+     * @param settings  the node-level settings
+     * @param name      the name of the executor
+     * @param size      the fixed number of threads
+     * @param queueSize the size of the backing queue
+     */
+    public PrioFixedExecutorBuilder(Settings settings, String name, int size, int queueSize) {
+        super(settings, name, size, queueSize);
+    }
+
+
+    @Override
+    ThreadPool.ExecutorHolder build(final Settings settings) {
+        final String nodeName = Node.NODE_NAME_SETTING.get(settings);
+        final int size = sizeSetting.get(settings);
+        final int queueSize = queueSizeSetting.get(settings);
+        final ThreadFactory threadFactory = EsExecutors.daemonThreadFactory(EsExecutors.threadName(nodeName, name()));
+        ExecutorService executor = new EsThreadPoolExecutor(
+            nodeName + "/" + name(),
+            size,
+            size,
+            0,
+            TimeUnit.MILLISECONDS,
+            new SizeBlockingQueue<>(new PriorityBlockingQueue<>(), queueSize),
```

Comment:
Wouldn't it be better to trade memory for performance by setting the `initialCapacity` to the queueSize and avoid any growing?